DatadogRUM/Sources/SDKMetrics/SessionEndedMetric.swift

```diff
@@ -81,29 +108,40 @@ internal struct SessionEndedMetric {
     ///   - sessionID: An ID of the session that is being tracked with this metric.
     ///   - precondition: The precondition that led to starting this session.
     ///   - context: The SDK context at the moment of starting this session.
+    ///   - tracksBackgroundEvents: If background events tracking is enabled for this session.
     init(
         sessionID: RUMUUID,
         precondition: RUMSessionPrecondition?,
-        context: DatadogContext
+        context: DatadogContext,
+        tracksBackgroundEvents: Bool
     ) {
         self.sessionID = sessionID
         self.bundleType = context.applicationBundleType
         self.precondition = precondition
+        self.tracksBackgroundEvents = tracksBackgroundEvents
+        self.ntpOffsetAtStart = context.serverTimeOffset
     }
 
     /// Tracks the view event that occurred during the session.
-    mutating func track(view: RUMViewEvent) throws {
+    /// - Parameters:
+    ///   - view: the view event to track
+    ///   - instrumentationType: the type of instrumentation used to start this view (only the first value for each `view.id` is tracked; succeeding values
+    ///   will be ignored so it is okay to pass value on first call and then follow with `nil` for next updates of given `view.id`)
+    mutating func track(view: RUMViewEvent, instrumentationType: ViewInstrumentationType?) throws {
```

Comment:
can this type be a reference type? 

struct with mutating methods are expensive and forces new copy of the object everytime we mutate.

Reply:
We started with reference type (`class SessionEndedMetric`), but [as discussed here](https://github.com/DataDog/dd-sdk-ios/pull/1866#discussion_r1627535883) the cost was on excessive locks for the sake of state synchronization. By its nature, updates to `SessionEndedMetric` will come from various places in the code and different threads. With that in mind we preferred simplicity in thread safety (single lock from `SessionEndedMetricController`) over memory footprint due to struct mutation.

To address memory footprint concerns:
- the SEM struct is optimized to not store large values - it pre-computes metric attributes
- the most frequent mutation (`track(view:)`) is linked to app lifecycle, and should be spaced out enough to be considered a rare event

Reply:
Makes sense, 

a followup to this how do we make sure the usage of the struct. If the updates are being called from different threads, a mutating struct is not thread safe.

Try yourself

```swift
    func testConcurrency() {
        var counter = Counter()

        DispatchQueue.concurrentPerform(iterations: 100) { _ in
            counter.increment()
        }

        DispatchQueue.concurrentPerform(iterations: 100) { _ in
            counter.decrement()
        }

        XCTAssertEqual(counter.value, 0)
    }

    struct Counter {
        var value = 0

        mutating func increment() {
            value += 1
        }

        mutating func decrement() {
            value -= 1
        }
    }
```

You have to call mutating methods from same queue (thread) with `sync` API if we want this to be thread safe mutating struct.

Reply:
Thread safety is enforced by design. The [`SessionEndedMetricController`](https://github.com/DataDog/dd-sdk-ios/blob/develop/DatadogRUM/Sources/SDKMetrics/SessionEndedMetricController.swift) is the only type that mutates the `SessionEndedMetric` struct. The controller is injected into codebase, making the struct a private detail. All other systems interact with the controller, not the `struct`.

By principle, the `struct SessionEndedMetric` should be made `private` to `SessionEndedMetricController` so it is not accessible in `internal` namespace. However, we keep it `internal` so it can be unit tested. Covering all edge cases for `SessionEndedMetric` mutation is trivial in `SessionEndedMetricTests`. Achieving the same through `SessionEndedMetricControllerTests` will visibly more hassle.

How does it sound @ganeshnj ?

Reply:
Also, `SEM` mutations through `SEMController` are tested for thread safety, see:

- [`func testTrackingSessionEndedMetricIsThreadSafe()`](https://github.com/DataDog/dd-sdk-ios/blob/07a23ef31523c19518b97f1cea6ba96c94f2ef9c/DatadogRUM/Tests/SDKMetrics/SessionEndedMetricControllerTests.swift#L93-L115)

Reply:
I see the lock on the [dictionary](https://github.com/DataDog/dd-sdk-ios/blob/07a23ef31523c19518b97f1cea6ba96c94f2ef9c/DatadogRUM/Sources/SDKMetrics/SessionEndedMetricController.swift#L14) which makes it thread safe.

Reply:
Thanks for clarifying this point on call 👍. The `SEM` is now a reference type to reduce memory footprint on mutations. Thread safety is still ensured by `SEMController`. After all it was a small but nice change, thanks!